SilverStripe 4 compatibility

robbieaverill opened the pull request:
## Note

Carries on from work done by @micmania1  in #405.

## Summary

Updated for compatibility with SilverStripe 4 in its current state. There are still some things to be done, but this is largely working to start off with. Brief summary of changes:

* Composer: dev alias added, PSR-4 autoloading compatibility
* Structure: "code" -> "src", namespacing added and implemented
* Travis: config updated for SS4 (PHP 7.x is no longer allowed to fail)
* PSR-2: some style formatting changes
* Asset building: Gulp used instead of compass

## Blockers

- [x] silverstripe/silverstripe-tagfield#87
- [x] silverstripe/silverstripe-lumberjack#48
- [x] silverstripe/silverstripe-comments#194
- [x] silverstripe/silverstripe-widgets#143
- [x] colymba/GridFieldBulkEditingTools#136

## Considerations

- [x] Do we need to create/update branches other than master before it becomes SS4 only?

dhensby commented:
OK - master is now aliased as `3.x-dev` and is where we should be putting SS 4 changes. please make sure to update `.travis.yml` and constraints in `composer.json`

robbieaverill commented:
@dhensby I've rebased and pushed. If you want me to squash a few of these commits, happy to do so

micmania1 commented:
Probably more a question for @sminnee / @chillu / @tractorcow but what's the guidance around committing the .upgrade.yml file? Is it useful after you've ran the upgrade tool?

@robbieaverill haven't tested this but looks good 👍 Glad to see those migration tasks gone :D

robbieaverill commented:
@micmania1 it's useful in that it tells other modules that are being upgraded what the namespaces to apply for Blog classes are. Without the mapping they'll all just be imported into the global namespace when you upgrade a module or project that uses this module, e.g. `use BlogPost;` as opposed to `use SilverStripe\Blog\Model\BlogPost;`

robbieaverill commented:
Once the next alpha release of framework comes out this PR or another will need to be updated to extend `PageController` instead of `Page_Controller`. Doing it now would probably fail builds by pulling in the latest alpha release

micmania1 commented on src/Model/BlogTag.php:
```diff
@@ -64,7 +64,7 @@ class BlogTag extends DataObject implements CategorisationObject
     protected function getListUrlSegment()
     {
-        return 'tags';
+        return 'tag';
```
Was this meant to be changed? It should match current functionality.

robbieaverill commented:
Yeah, this changes it back to current functionality
On Fri, 13 Jan 2017 at 4:01 PM, Michael Strong <notifications@github.com>
wrote:

> *@micmania1* commented on this pull request.
>
>
>
>
> ------------------------------
>
>
>
>
> In src/Model/BlogTag.php
> <https://github.com/silverstripe/silverstripe-blog/pull/421#pullrequestreview-16502018>
> :
>
>
> > @@ -64,7 +64,7 @@ class BlogTag extends DataObject implements CategorisationObject
>
>       */
>
>      protected function getListUrlSegment()
>
>      {
>
> -        return 'tags';
>
> +        return 'tag';
>
>
>
> Was this meant to be changed? It should match current functionality.
>
>
>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/silverstripe/silverstripe-blog/pull/421#pullrequestreview-16502018>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE7lnlmVIrGQut4hSXHCpDisTXYjzQ-pks5rRukCgaJpZM4LRcdk>
> .
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>

dhensby commented:
Looks like the comments module needs upgrading?

robbieaverill commented:
Yeah it does, and there is some widget work to do too. I'll be looking at this early next week.

dhensby commented:
Ok, then they are another couple of blockers to list

robbieaverill commented:
Recent discussion - should we move the comments and widgets code into a separate module and suggest that instead, e.g. `silverstripe/blog-comments` and `silverstripe/blog-widgets` ?

sminnee commented:
> Recent discussion - should we move the comments and widgets code into a separate module and suggest that instead, e.g. silverstripe/blog-comments and silverstripe/blog-widgets ?

I'd suggest getting an SS4 version merged first and then look at further refactorings. Otherwise this will becoming a zombie PR.

robbieaverill commented:
@sminnee I just reached for the zombie emoji response - sadly disappointed.

tractorcow commented:
Testing this locally:
 
 - [ ] categories and tags aren't saving

tractorcow commented:
There's a test failing; I'll push a fix if you give me write access. :)

robbieaverill commented:
@tractorcow I've fixed the tests, I had to update a few things:

* There were parts where I'd missed namespacing the fixture classes in the tests, and some parts (blog widget classes) where I seem to have completely forgotten to implement namespaces at all
* [This multibyte logic](https://github.com/silverstripe/silverstripe-blog/blob/2.4.5/code/extensions/URLSegmentExtension.php#L39-L43) got lost somehow, re-introduced
* [This logic](https://github.com/silverstripe/silverstripe-blog/blob/2.4.5/code/extensions/URLSegmentExtension.php#L23) was changed to `if (empty($this->URLSegment))` during the early refactor and move the the BlogObject trait. I've added the empty check back, checking second for the URL Segment being empty - this matches existing behaviour (v2.4.5)
* I've switched out other places where FQCNs were string literals for `BlogPost::class` equivalents with imports at the top
* I added namespaced to the tests while I was in there

---

Another thing I noticed was that lumberjack isn't removing blog posts from the SiteTree when they're displayed in the GridField. Logged as silverstripe/silverstripe-lumberjack#50

robbieaverill commented:
@dhensby - we're green, finally 💚

dhensby commented:
43 mins is still far too long ;)
